daemon/api_sideload_n_try.go

```diff
@@ -123,38 +125,119 @@ func sideloadOrTrySnap(c *Command, body io.ReadCloser, boundary string) Response
 		}
 		return trySnap(c.d.overlord.State(), form.Values["snap-path"][0], flags)
 	}
-	flags.RemoveSnapPath = true
 
+	flags.RemoveSnapPath = true
 	flags.Unaliased = isTrue(form, "unaliased")
 	flags.IgnoreRunning = isTrue(form, "ignore-running")
-	systemRestartImmediate := isTrue(form, "system-restart-immediate")
 
-	origPath, tempPath, errRsp := form.SnapFileNameAndTempPath()
+	origPaths, tempPaths, errRsp := form.GetSnapFiles()
 	if errRsp != nil {
 		return errRsp
 	}
 
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var chg *state.Change
+	if len(origPaths) > 1 {
+		chg, errRsp = sideloadManySnaps(st, form, origPaths, tempPaths, flags, user)
+	} else {
+		chg, errRsp = sideloadSnap(st, form, origPaths[0], tempPaths[0], flags)
+	}
+	if errRsp != nil {
+		return errRsp
+	}
+
+	systemRestartImmediate := isTrue(form, "system-restart-immediate")
+	if systemRestartImmediate {
+		chg.Set("system-restart-immediate", true)
+	}
+
+	ensureStateSoon(st)
+
+	// the handoff is only done when the unlock succeeds (instead of panicking)
+	// but this is good enough
+	pathsToNotRemove = tempPaths
+
+	return AsyncResponse(nil, chg.ID())
+}
+
+func sideloadManySnaps(st *state.State, form *Form, origPaths, tempPaths []string, flags snapstate.Flags, user *auth.UserState) (*state.Change, *apiError) {
+	var sideInfos []*snap.SideInfo
+	var names []string
+
+	for i, tempPath := range tempPaths {
+		si, apiError := readSideInfo(st, form, tempPath, origPaths[i])
+		if apiError != nil {
+			return nil, apiError
+		}
+
+		sideInfos = append(sideInfos, si)
+		names = append(names, si.RealName)
+	}
+
+	var userID int
+	if user != nil {
+		userID = user.ID
+	}
+
+	tss, err := snapstateInstallPathMany(context.TODO(), st, sideInfos, tempPaths, userID, &flags)
+	if err != nil {
+		return nil, errToResponse(err, tempPaths, InternalError, "cannot install snap files: %v")
+	}
+
+	msg := fmt.Sprintf(i18n.G("Install snaps %q from files %q"), names, origPaths)
+	chg := newChange(st, "install-snaps", msg, tss, names)
+	chg.Set("api-data", map[string][]string{"snap-names": names})
+
+	return chg, nil
+}
+
+func sideloadSnap(st *state.State, form *Form, origPath, tempPath string, flags snapstate.Flags) (*state.Change, *apiError) {
 	var instanceName string
 	if len(form.Values["name"]) > 0 {
```

Comment:
this can then check `sideloadSnap.InstanceName != ""` and form is not needed anymore